src/examples/libpmem2/unsafe_shutdowns.c

```diff
@@ -0,0 +1,429 @@
+ *       notice, this list of conditions and the following disclaimer.
+ *
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in
+ *       the documentation and/or other materials provided with the
+ *       distribution.
+ *
+ *     * Neither the name of the copyright holder nor the names of its
+ *       contributors may be used to endorse or promote products derived
+ *       from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * unsafe_shutdowns.c -- unsafe shutdowns example for the libpmem2
+ */
+
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <libpmem2.h>
+
+#ifndef _WIN32
+#include <unistd.h>
+#else
+#include <io.h>
+#endif
+
+/* flags for struct shutdown_state */
+#define CLEAR_FLAG 0
+#define DIRTY_FLAG 1
+
+/*
+ * SDS_LEN -- overall struct shutdown_state len
+ */
+#define SDS_LEN 4096
+
+/*
+ * UUID_LEN -- length of the struct shutdown_state without fields: usc,
+ * flag, checksum (SDS_LEN - 8 - 1 - 8 = 4079)
+ */
+#define UUID_LEN 4079
+
+/*
+ * SDS_LEN_NO_CS -- length of the struct shutdown_state without checksum
+ * field
+ */
+#define SDS_LEN_NO_CS offsetof(struct shutdown_state, checksum)
+
+/*
+ * shutdown_state -- structure holding data about shutdown state
+ */
+struct shutdown_state {
+	uint64_t usc; /* unsafe shutdown count */
+	uint8_t uuid[UUID_LEN]; /* UID */
+	uint8_t flag;
+	uint64_t checksum;
+};
+
+/*
+ * checksum_compute -- calculate checksum of the structure using
+ * Fletcher's 64 algorithm
+ */
+static void
+checksum_compute(void *addr, size_t len, uint64_t *csump)
+{
+	if (len % 4 != 0)
+		abort();
+
+	uint32_t *p32 = addr;
+	uint32_t *p32end = (uint32_t *)((char *)addr + len);
+	uint32_t lo32 = 0;
+	uint32_t hi32 = 0;
+
+	while (p32 < p32end) {
+			lo32 = (lo32 + *p32) % 255;
+			hi32 = (hi32 + lo32) % 255;
+			++p32;
+	}
+
+	*csump = (uint64_t)hi32 << 32 | lo32;
+}
+
+/*
+ * prepare_environment -- prepare environment for create/read operation
+ */
+static void
+prepare_environment(int *fd, char *file, struct pmem2_config *cfg,
```

Comment:
Why do you need this function? Move the common logic in front of the divergent path. That way you won't need a function that takes gazillion parameters and disrupts the flow.